src/meta/src/rpc/service/notification_service.rs

```diff
@@ -59,116 +61,172 @@ where
-    async fn subscribe(
-        &self,
-        request: Request<SubscribeRequest>,
-    ) -> Result<Response<Self::SubscribeStream>, Status> {
-        let req = request.into_inner();
-        let subscribe_type = req.get_subscribe_type()?;
-        if subscribe_type == SubscribeType::Frontend {
-            self.hummock_manager.pin_snapshot(req.worker_id).await?;
-        }
-        let host_address = req.get_host()?.clone();
-
-        let (tx, rx) = mpsc::unbounded_channel();
 
+    async fn get_catalog_snapshot(&self) -> (Catalog, Vec<UserInfo>, NotificationVersion) {
         let catalog_guard = self.catalog_manager.get_catalog_core_guard().await;
-        let (databases, schemas, mut tables, sources, sinks, indexes, views) =
+        let (databases, schemas, tables, sources, sinks, indexes, views) =
             catalog_guard.database.get_catalog();
-        let creating_tables = catalog_guard.database.list_creating_tables();
         let users = catalog_guard.user.list_users();
+        let notification_version = self.env.notification_manager().current_version().await;
+        (
+            (databases, schemas, tables, sources, sinks, indexes, views),
+            users,
+            notification_version,
+        )
+    }
 
+    async fn get_parallel_unit_mapping_snapshot(
+        &self,
+    ) -> (Vec<ParallelUnitMapping>, NotificationVersion) {
         let fragment_guard = self.fragment_manager.get_fragment_read_guard().await;
         let parallel_unit_mappings = fragment_guard.all_fragment_mappings().collect_vec();
-        let all_internal_tables = fragment_guard.all_internal_tables();
-        let hummock_snapshot = Some(self.hummock_manager.get_last_epoch().unwrap());
+        let notification_version = self.env.notification_manager().current_version().await;
+        (parallel_unit_mappings, notification_version)
+    }
 
-        // We should only pin for workers to which we send a `meta_snapshot` that includes
-        // `HummockVersion` below. As a result, these workers will eventually unpin.
-        if subscribe_type == SubscribeType::Hummock {
-            self.hummock_manager
-                .pin_version(req.get_worker_id())
-                .await?;
-        }
+    async fn get_worker_node_snapshot(&self) -> (Vec<WorkerNode>, NotificationVersion) {
+        let cluster_guard = self.cluster_manager.get_cluster_core_guard().await;
+        let nodes = cluster_guard.list_worker_node(WorkerType::ComputeNode, Some(Running));
+        let notification_version = self.env.notification_manager().current_version().await;
+        (nodes, notification_version)
+    }
 
+    async fn get_hummock_version_snapshot(&self) -> (HummockVersion, NotificationVersion) {
         let hummock_manager_guard = self.hummock_manager.get_read_guard().await;
+        let hummock_version = hummock_manager_guard.current_version.clone();
+        let notification_version = self.env.notification_manager().current_version().await;
```

Comment:
It seems that `NotificationManager::current_version` would acquire a lock which may hold by other thread for `NotificationManagerCore::notify`, this method would call `increase_version` and cause an IO operation on `MetaStore` (ETCD) and hold this lock for a long time. So I suggest releasing read-guard before calling `current_version`

Reply:
We can't drop `hummock_manager_guard` before we get the notification version. If we do that, between line 96 and line 97 other thread may modify hummock and increase notification version. Then the `notification_version` in line 97 can't represent the order of `hummock_version`(line 96) in the multiple notifications. It will let us treat the other thread's modification as a notification which is sent before we get `hummock_version`. We will loss the notification of this modification.